src/cargo/sources/registry/mod.rs

```diff
@@ -790,7 +796,13 @@ impl<'gctx> Source for RegistrySource<'gctx> {
                 .index
                 .query_inner(dep.package_name(), &req, &mut *self.ops, &mut |s| {
                     let matched = match kind {
-                        QueryKind::Exact => dep.matches(s.as_summary()),
+                        QueryKind::Exact => {
+                            if self.gctx.cli_unstable().unstable_options {
```

Comment:
Same here. We might want this check to prevent any regression in the future, if some underlying functions change.

```suggestion
                            if req.is_precise() && self.gctx.cli_unstable().unstable_options {
```